### ISSUE TITLE
Clean up asciimaps

### ISSUE THREAD

**Issue by youngmit** [terrapower/armi#436]
The functionality in utils/asciimaps.py has been pretty brittle, and difficult to extend and debug, and frankly pretty error prone to use. We've been talking for a while about perhaps taking another stab at the problem, so I am making this ticket to get it out there as a thing to do. I think a lot of the complexity stems from modelling the problem with a bunch of classes, which have pretty abstract state that needs to be ushered through a handful of methods that are subclassed in pretty nuanced and difficult to understand ways. I think the problem of mapping from a structured text grid to a dictionary of i,j indices -> specifier would better be suited to straight-up functions (e.g., `fullHexFlatsUpToAscii(d)` and its inverse `fullHexFlatsUpToDict(text)`. We can still provide a sort of pseudo-factory, which returns the appropriate functions given a provided geom and symmetry condition. If whoever tackles this feels like the use of some private classes would ease implementation of these functions, they are more than welcome to do so, or even try using the classes that we already have to get a fist-cut implementation of the functional interface, but I suspect that the class approach is more complicated than it needs to be.

In summary:
 - replace the classes in asciimaps.py with functions for going to/from ascii<->dict
 - refactor client code to use the functions. life should be way simpler!

**Comment by youngmit** [comment]
@davis8dd think you'd be interested?

**Comment by scottyak** [comment]
Question -- I'm considering using [dataclasses](https://docs.python.org/3.7/library/dataclasses.html) as an implementation detail (since [`__post_init__`](https://docs.python.org/3.7/library/dataclasses.html#post-init-processing) is nice), but it is only available from 3.7 onwards. I suspect that's okay, since the latest release of ARMI appears to be tested against 3.7, 3.8, and 3.9 but not 3.6: https://github.com/terrapower/armi/runs/4267083946?check_suite_focus=true

However, since setup.py says `>=3.6`, I thought it would be good to check first. (there's a typo there too -- "s/requres/requires")
https://github.com/terrapower/armi/blob/b94744fdf12f247828dd07bf99e251651b494a13/setup.py#L49


So can I assume the user has python `>=3.7` ?

**Comment by jakehader** [comment]
I think it's a fair assumption that most users have Python 3.7+ internally at TerraPower, but I cannot speak for all external users. Right now the installation docs do state Python 3.6+ even though the setup.py is broken.

**Comment by scottyak** [comment]
It seems, then, that either release v0.2 should be tested on python 3.6 (since it seems like it currently isn't), or that setup.py and the release doc should be updated to say it requires >=3.7 ?

I think dropping python 3.6 support might not be a terrible idea -- if someone is using python 3.6, they could depend on the previous release. Since release v0.2 was only 5 days ago, it seems unlikely that anyone is relying on it yet. This could be a good opportunity to drop python 3.6 support -- otherwise you might have to wait till the next major release to do so :(

The reason why I'm hoping that python 3.6 support can be dropped is that, besides dataclasses, python 3.7 also brings [types and generics](https://www.python.org/dev/peps/pep-0560/), which would make refactorings and cleanups so much easier. Perhaps some subclasses of `abc.ABC` can also be simplified.

**Comment by jakehader** [comment]
@john-science - maybe we can hot patch the v0.2 release to enforce 3.7+? I don't see a reason why we need to support 3.6 at this time.

**Comment by john-science** [comment]
@jakehader @scottyak  My preference is to only support Python 3.7-3.9, as those are really the only versions that we test for seriously. And while I'm generally a fan of supporting a wide breadth of versions for an importable library... I don't know, I think for a scientific modeling framework I'm more in favor of constraining the versions to be better tested.

Let me ask around, and see if anyone depends on 3.6.

**Comment by john-science** [comment]
@jakehader @scottyak  It looks like no one is using Python 3.6 anyway.  Nick is hesitant to drop support for 3.10, so I _think_ saying we support Python 3.7+ is a pretty safe bet for everyone.

**Comment by jakehader** [comment]
@scottyak, the 3.6 requirement has been removed in #491.

**Comment by john-science** [comment]
@sdow1ll Wants to take this ticket.

**Comment by davis8dd** [comment]
Sorry for leaving this one assigned for so long!

**Comment by sdow1ll** [comment]
Hi @john-science I have a working Jupyter Notebook of the asciimaps.py file that basically has the converted functions from the asciimap class, but one thing that I got stuck on was that I noticed at the end of the asciimaps.py file, there is a code block of the following:

`def asciiMapFromGeomAndSym(geomType: str, symmetry: str):
    """Get a ascii map class from a geometry type."""
    from armi.reactor import geometry

    MAP_FROM_GEOM = {
        (geometry.HEX, geometry.THIRD_CORE): AsciiMapHexThird,
        (geometry.HEX, geometry.FULL_CORE): AsciiMapHexFullTipsUp,
        (geometry.CARTESIAN, None): AsciiMapCartesian,
        (geometry.CARTESIAN, geometry.FULL_CORE): AsciiMapCartesian,
    }

    return MAP_FROM_GEOM[(geomType, symmetry)]`

Is this something that I should try to edit as well? When looking at this, I notice that the MAP_FROM_GEOM dictionary relies on the original classes. Are the functions I wrote basically an additional feature to this file, and that I should not touch the following code block above?

**Comment by john-science** [comment]
The only two references to that function are here:

https://github.com/terrapower/armi/blob/6a546dd1ca3a81128e5dcc469a5d49853db2047c/armi/reactor/blueprints/gridBlueprint.py#L410

https://github.com/terrapower/armi/blob/6a546dd1ca3a81128e5dcc469a5d49853db2047c/armi/reactor/blueprints/gridBlueprint.py#L567

So, refactoring those two use-cases will be necessary to do your refactoring of the asciimaps utils.

(Sorry for the long delay.)
